src/runtime-rs/crates/kata/src/lib.rs

```diff
@@ -0,0 +1,318 @@
+    /// the id of the container
+    pub id: String,
+    /// the namespace for the container
+    pub namespace: String,
+    /// the address of the containerd's main socket
+    pub address: String,
+    /// the binary path to publish events back to containerd
+    pub publish_binary: String,
+    /// the path to the bundle to delete
+    pub bundle: String,
+    /// Whether or not to enable debug
+    pub debug: bool,
+}
+
+impl ShimArgs {
+    /// Check the shim argument object is vaild or not.
+    ///
+    /// The id, namespace, address and publish_binary are mandatory for START, RUN and DELETE.
+    /// And bundle is mandatory for DELETE.
+    pub fn validate(&mut self, is_delete: bool) -> Result<()> {
+        if self.id.is_empty()
+            || self.namespace.is_empty()
+            || self.address.is_empty()
+            || self.publish_binary.is_empty()
+        {
+            return Err(Error::InvalidArgument);
+        }
+        if is_delete && self.bundle.is_empty() {
+            return Err(Error::InvalidArgument);
+        }
+
+        if Self::is_component_dangerous(&self.id) || Self::is_component_dangerous(&self.namespace) {
+            return Err(Error::InvalidArgument);
+        }
+
+        // Ensure `address` is a valid path.
+        let path = PathBuf::from(self.address.clone())
+            .canonicalize()
+            .map_err(|_| Error::InvalidArgument)?;
+        let md = path.metadata().map_err(|_| Error::InvalidArgument)?;
+        if !md.file_type().is_socket() {
+            return Err(Error::InvalidArgument);
+        }
+        self.address = path
+            .to_str()
+            .map(|v| v.to_owned())
+            .ok_or(Error::InvalidArgument)?;
+
+        // Ensure `bundle` is a valid path.
+        if !self.bundle.is_empty() {
+            let path = PathBuf::from(self.bundle.clone())
+                .canonicalize()
+                .map_err(|_| Error::InvalidArgument)?;
+            let md = path.metadata().map_err(|_| Error::InvalidArgument)?;
+            if !md.is_dir() {
+                return Err(Error::InvalidArgument);
+            }
+            self.bundle = path
+                .to_str()
+                .map(|v| v.to_owned())
+                .ok_or(Error::InvalidArgument)?;
+        }
+
+        Ok(())
+    }
+
+    fn is_component_dangerous(comp: &str) -> bool {
+        if comp.is_empty() {
+            return true;
+        }
+        // only allow ascii alphanumeric character and '-', '_', '.' and '~'
+        !comp
+            .chars()
+            .all(|x| matches!(x, '0'..='9' | 'A'..='Z' | 'a'..='z' | '-' | '_' | '.' | '~'))
+    }
+}
+
+/// Command executor for shim.
+pub struct ShimExecutor {
+    args: ShimArgs,
+}
+
+impl ShimExecutor {
+    /// Create a new instance of [`Shim`].
+    pub fn new(args: ShimArgs) -> Self {
+        ShimExecutor { args }
+    }
+
+    fn get_bundle_path(&self) -> Result<PathBuf> {
+        std::env::current_dir().map_err(Error::BundlePath)
```

Comment:
Shouldn't this function prefer to return `self.args.bundle` and only fall back to `current_dir()` if `self.args.bundle` is empty?